### DISCUSSION TITLE
useForm getInputProps return type

### DISCUSSION THREAD

**Discussion by gdostie** [mantinedev/mantine#2306] [Feature requests]
`getInputProps` function exposed on the `useForm` hook returning object has its return type specified as `any` ([code here](https://github.com/mantinedev/mantine/blob/master/src/mantine-form/src/types.ts#L59-L62)). However the documentation website already states this:

> getInputProps – returns an object with value, onChange and error that should be spread on input

This information could be reflected directly in the type declaration. Although not perfect, because the `value` remains typed as `unknown`, something like the following would be a step in the right direction:

```ts
type SharedInputProps = {
  onFocus?(): void;
  error?: FormErrors;
}

type InputProps = SharedInputProps & { value: unknown; onChange(value: unknown): void; };

type CheckboxProps = SharedInputProps & { checked: boolean; onChange(checked: boolean): void; };

type GetInputPropsOptions = {
  type?: GetInputPropsType;
  withError?: boolean;
  withFocus?: boolean;
}

export type GetInputProps<Values> = <Field extends LooseKeys<Values>, Options extends GetInputPropsOptions>(
  path: Field,
  options?: Options
) => Options['type'] extends 'checkbox' ? CheckboxProps : InputProps;
```

[Try it in the TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzjAnmApnASughgYxgDkIATTAXzgDMoIQ4ByKPQxgbgChPUM4BxdDACSAOzABXGAAU6YAM4AVNJgC8TYOKmM4AHyb4AFunwBrAEYQAHh269MAGQgR56ANLoU8gDwA1XAA2EujyAHxw6qaeENRw-kEhenAAFPIwUJoA5nAAZIgUAJRcPCpwAGLQIACiUHRQ8hHYJtCk3mkZopkANE0ExGToocX2cADKhrgspGKSMnIN6giccHAQohX4EvIA-MkFAFxwAG4QwKRcK+i10NuHFVDV1-VcFHalM1KyEAqN45Po0y0c2+DTySCOgWChwkolMoggAHdROxVqIAMITTroZIQhLQ2HwpEHY6nc5wCjDUoYkwWaxfH7qP5TD7An5guBGGkAw6WCABPDI1EY3BY5KcszcuC8-ki4knM4oilvPiCERA+nyADyYBgwDWC0Qy2QKluAiELI1ygwFzgCOAMEMNTqpulAptdodGy2LucMuRnFenHQVkgsGNKvN6vmfkhIXC6m8ZWA6ACpDgwZg6FEpAaThc7k8PniwTCPW1uv16asmezDVVFvm5b1ojCySNYFwDruydTXSN3wrLdNTf1nAKEXCI5bAG1GPZGABdKs1nMGYxmSw2ODbODUjd0+ZwQ4NkHFTgAenPyBCMHkyswAFkUPcGIsaFAJPb5IckKJcCB0EOdoshRAJgCiUgeV9AVyWnJdA3IfAAn+Dl9XgTJI1mDVDnrKMQW8J8XyGbh8DQuAwEPdQMLVLD5mSRhaE-W9GCKC8rwokEADoSLIzRaJBRpqJPBR6MYr9OIABk4v8AMYHokHsQ5GD47RyVYy84BU1l5G4zhSJbeBxVpKwNUEzDPjohiP3EqSwIguTEHDQC1xpTcdEKLgNKMzcNW4oA)

**Comment by rtivital** [comment]
It's very difficult to add correct type to `form.getInputProps` because it is supposed to be used with every input component + native inputs. We previously had this type (similar to what you've proposed), but we had a lot of reports of Discord that it does not work well in some edge cases. `any` is doing good enough in this case, since return value is supposed to be spread on the input – you should not use it any other way.

**Comment by gdostie** [comment]
Ahh okay I see, makes sense.

> you should not use it any other way

I am not, I was building a custom form component that Mantine doesn't have (a CodeEditor built on top of monaco editor), and I want this component to be used in the same way as the other form components. It would have been useful to know what this function returns so that my component can support it, but in the end I was able to figure it out by looking at the source code.

**Comment by GermainBergeron** [comment]
Spreading the prop on a component does cause issue though. The current return type is making us lose type safety. I've made a small [codesandbox](https://codesandbox.io/s/mantine-forked-y03g2r?file=/src/App.tsx) to demonstrate the issue

<img width="741" alt="image" src="https://user-images.githubusercontent.com/260007/187976994-794e4f79-6c2c-4f50-a130-0435134e3a2b.png">

**Reply by GermainBergeron** [reply]
It seems like using `Record<string, unknown>` or something similar would be better than `any` as it keeps the type safety
